jobs/jobs.go

```diff
@@ -0,0 +1,326 @@
+	OnError(func(error))
+	// OnJobChange sets handler for job update.
+	OnJobChange(func(map[JobExecutionState][]JobExecutionSummary))
+	// GetPendingJobs gets list of pending jobs.
+	GetPendingJobs(ctx context.Context) (map[JobExecutionState][]JobExecutionSummary, error)
+	// DescribeJob gets details of specific job.
+	DescribeJob(ctx context.Context, id string) (*JobExecution, error)
+	// UpdateJob updates job status.
+	UpdateJob(ctx context.Context, j *JobExecution, s JobExecutionState, opt ...UpdateJobOption) error
+}
+
+type jobs struct {
+	mqtt.ServeMux
+	cli         mqtt.Client
+	thingName   string
+	mu          sync.Mutex
+	chResps     map[string]chan interface{}
+	onError     func(err error)
+	onJobChange func(map[JobExecutionState][]JobExecutionSummary)
+	msgToken    int
+}
+
+func (j *jobs) token() string {
+	j.msgToken++
+	return fmt.Sprintf("%x", j.msgToken)
+}
+
+func (j *jobs) topic(operation string) string {
+	return "$aws/things/" + j.thingName + "/jobs/" + operation
+}
+
+// New creates IoT Jobs interface.
+func New(ctx context.Context, cli awsiotdev.Device) (Jobs, error) {
+	j := &jobs{
+		cli:       cli,
+		thingName: cli.ThingName(),
+		chResps:   make(map[string]chan interface{}),
+	}
+	for _, sub := range []struct {
+		topic   string
+		handler mqtt.Handler
+	}{
+		{j.topic("notify"), mqtt.HandlerFunc(j.notify)},
+		{j.topic("+/get/accepted"), mqtt.HandlerFunc(j.getJobAccepted)},
+		{j.topic("+/get/rejected"), mqtt.HandlerFunc(j.rejected)},
+		{j.topic("+/update/accepted"), mqtt.HandlerFunc(j.updateJobAccepted)},
+		{j.topic("+/update/rejected"), mqtt.HandlerFunc(j.rejected)},
+		{j.topic("get/accepted"), mqtt.HandlerFunc(j.getAccepted)},
+		{j.topic("get/rejected"), mqtt.HandlerFunc(j.rejected)},
+	} {
+		if err := j.ServeMux.Handle(sub.topic, sub.handler); err != nil {
+			return nil, err
+		}
+	}
+
+	err := cli.Subscribe(ctx,
+		mqtt.Subscription{Topic: j.topic("notify"), QoS: mqtt.QoS1},
+		mqtt.Subscription{Topic: j.topic("get/#"), QoS: mqtt.QoS1},
+		mqtt.Subscription{Topic: j.topic("+/get/#"), QoS: mqtt.QoS1},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return j, nil
+}
+
+func (j *jobs) notify(msg *mqtt.Message) {
+	m := &jobExecutionsChangedMessage{}
+	if err := json.Unmarshal(msg.Payload, m); err != nil {
+		j.handleError(err)
+		return
+	}
+	j.mu.Lock()
+	cb := j.onJobChange
+	j.mu.Unlock()
+
+	if cb != nil {
+		go cb(m.Jobs)
+	}
+}
+
+func (j *jobs) GetPendingJobs(ctx context.Context) (map[JobExecutionState][]JobExecutionSummary, error) {
+	req := &simpleRequest{ClientToken: j.token()}
+	ch := make(chan interface{}, 1)
+	j.mu.Lock()
+	j.chResps[req.ClientToken] = ch
+	j.mu.Unlock()
+	defer func() {
+		j.mu.Lock()
+		delete(j.chResps, req.ClientToken)
+		j.mu.Unlock()
+	}()
+
+	breq, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	if err := j.cli.Publish(ctx,
+		&mqtt.Message{
+			Topic:   j.topic("get"),
+			QoS:     mqtt.QoS1,
+			Payload: []byte(breq),
```

Comment:
The type casting is not necessary since `breq` is `[]byte`. (other breqs are as well)